lib/puppet/configurer.rb

```diff
@@ -340,6 +340,14 @@ def run_internal(options)
         rescue StandardError => detail
           Puppet.warning(_("Unable to fetch my node definition, but the agent run will continue:"))
           Puppet.warning(detail)
+
+          unless configured_environment
+            Puppet.debug(_("Failed to fetch node definition, attempting to find out the last used environment"))
+            if (env = last_used_environment)
+              @environment = env
+              report.environment = @environment
+            end
+          end
```

Comment:
I was hoping to eliminate the node request entirely and always use the environment from:

1. environment from the `cached_catalog` (when running with `puppet agent --use_cached_catalog)
2. environment from the catalog passed in as `options[:catalog]` (the `puppet apply` case)
3. environment from last run summary (if it exists)
4. `Puppet[:environment]`

Reply:
I removed the node request and updated the tests (didn't write tests for the new behavior yet).

The environment selection logic remains the same as before, with the caveat that `last_run_environment` is checked if no (cached) catalog is used _and_ `Puppet[:environment]` is not set via config or CLI.